chronos/asyncsync.nim

```diff
@@ -180,6 +207,55 @@ proc isSet*(event: AsyncEvent): bool =
   ## Return `true` if and only if the internal flag of ``event`` is `true`.
   event.flag
 
+proc wakeupNext(waiters: var seq[Future[void]]) {.inline.} =
+  var i = 0
+  while i < len(waiters):
+    var waiter = waiters[i]
+    inc(i)
+
+    if not(waiter.finished()):
+      waiter.complete()
+      break
+
+  if i > 0:
+    waiters.delete(0, i - 1)
+
+proc newAsyncSemaphore*(value: int = 1): AsyncSemaphore =
+  ## Creates a new asynchronous bounded semaphore ``AsyncSemaphore`` with
+  ## internal counter set to ``value``.
+  doAssert(value >= 0, "AsyncSemaphore initial value must be bigger or equal 0")
+  discard getThreadDispatcher()
+  AsyncSemaphore(waiters: newSeq[Future[void]](), counter: value,
+                 maxcounter: value)
+
+proc locked*(asem: AsyncSemaphore): bool =
+  ## Returns ``true`` if semaphore can not be acquired immediately
+  (asem.counter == 0)
+
+proc acquire*(asem: AsyncSemaphore) {.async.} =
+  ## Acquire a semaphore.
+  ##
+  ## If the internal counter is larger than zero on entry, decrement it by one
+  ## and return immediately. If its zero on entry, block and wait until some
+  ## other task has called ``release()`` to make it larger than 0.
+  while asem.counter <= 0:
```

Comment:
the loop should not be needed here - there will be exactly one queue item per count below zero so they should line up perfectly